Rubberduck.Core/UI/CodeExplorer/Commands/ImportCommand.cs

```diff
@@ -100,43 +116,94 @@ protected override void OnExecute(object parameter)
-                    {
-                        project?.Dispose();
-                    }
-                    return;
+                    return new List<string>();
                 }
 
-                var fileExists = dialog.FileNames.Select(s => s.Split('.').Last());
-                if (fileExists.Any(fileExt => !_importableExtensions.Contains(fileExt)))
+                var fileNames = dialog.FileNames;
+                var fileExtensions = fileNames.Select(Path.GetExtension);
+                if (fileExtensions.Any(fileExt => !_importableExtensions.Contains(fileExt)))
                 {
-                    if (usingFreshProjectWrapper)
-                    {
-                        project.Dispose();
-                    }
-                    return;
+                    NotifyUserAboutAbortDueToUnsupportedFileExtensions(fileNames);
+                    return new List<string>();
                 }
 
-                foreach (var filename in dialog.FileNames)
+                return fileNames;
+            }
+        }
+
+        protected virtual string DialogsTitle => RubberduckUI.ImportCommand_OpenDialog_Title;
+
+        private void NotifyUserAboutAbortDueToUnsupportedFileExtensions(IEnumerable<string> fileNames)
+        {
+            var firstUnsupportedFile = fileNames.First(filename => !_importableExtensions.Contains(Path.GetExtension(filename)));
+            var unsupportedFileName = Path.GetFileName(firstUnsupportedFile);
+            var message = string.Format(RubberduckUI.ImportCommand_UnsupportedFileExtensions, unsupportedFileName);
+            MessageBox.NotifyWarn(message, DialogsTitle);
+        }
+
+        private void ImportFilesWithSuspension(ICollection<string> filesToImport, IVBProject targetProject)
+        {
+            var suspensionResult = _parseManager.OnSuspendParser(this, new[] {ParserState.Ready}, () => ImportFiles(filesToImport, targetProject));
+            if (suspensionResult != SuspensionResult.Completed)
+            {
+                Logger.Warn("File import failed due to suspension failure.");
+            }
+        }
+
+        protected virtual void ImportFiles(ICollection<string> filesToImport, IVBProject targetProject)
+        {
+            using (var components = targetProject.VBComponents)
+            {
+                foreach (var filename in filesToImport)
                 {
-                    using (var components = project.VBComponents)
+                    var fileExtension = Path.GetExtension(filename);
+                    if (fileExtension != null
+                        && ComponentTypeForExtension.TryGetValue(fileExtension, out var componentType)
+                        && componentType == ComponentType.Document)
                     {
-                        components.Import(filename);
+                        //We have to dispose the return value.
+                        using (components.ImportSourceFile(filename)) { }
```

Comment:
if this throws, we're bailing out of the file loop. is this intended?

Reply:
I was not aware this can throw an expected exception.